### PR TITLE
Use QlobberSub (store clientId -> topic -> qos)

### PR DESCRIPTION
**Pull request by davedoesdev**
`QlobberSub` is copied from https://github.com/mcollina/aedes-persistence/pull/18 so should probably be shared. @mcollina what's your preferred way to do that?

@behrad this _should_ speed things up.

Are there any benchmarks for this module alone?

### CODE REVIEW

**Comment by behrad** [comment]
@davedoesdev Great effort, This helped me to create 2 comparable branches of aedes-cached-persistence:

1. Porting your QosQlobber: https://github.com/behrad/aedes-cached-persistence/blob/QosQlobber/index.js (haven't compared that code with this PR yet)

2. Using Dedup with stringified values: https://github.com/behrad/aedes-cached-persistence/blob/DedupQlobber/index.js

I'm going to send respective results.

**Comment by behrad** [comment]
| fork | offlineClients | subscriptions | Startup Time | Heap/RSS |
-------|------|-------|------|-------|
|qosQlobber|1439630|3751115|245962.766ms|6.9G/7.8G
|DedupQlobber|1439559|3751201|207796.314ms|5.9G/6.7G

- The above branches are draft implementations to benchmark both directions.

- DedupQlobber branch has some un-answered questions yet like https://github.com/mcollina/aedes-persistence/issues/19

**Comment by behrad** [comment]
I have not tested publishing qos>=1 (i.e. topic matching) yet, @davedoesdev can you please elaborate `.match` speed between the two? 
How much will they differ?

**Comment by davedoesdev** [comment]
@behrad how is the QlobberDedup version working?

Also, do you need this line? https://github.com/behrad/aedes-cached-persistence/blob/QosQlobber/index.js#L45

**Comment by behrad** [comment]
> how is the QlobberDedup version implemented?
https://github.com/behrad/aedes-cached-persistence/blob/DedupQlobber/index.js

This is simply storing concatenated value strings (clientId;topic;qos) inside QlobberDedup

**Comment by behrad** [comment]
> Also, do you need this line?

if QosQlobber is deduping based on values, No!

**Comment by behrad** [comment]
Can https://github.com/chad3814/node-hashtable  help reducing memory usage @davedoesdev ?

**Comment by davedoesdev** [comment]
@behrad how is DedupQlobber dedup'ing? It seems to be using normal Qlobber.

**Comment by behrad** [comment]
> how is DedupQlobber dedup'ing? It seems to be using normal Qlobber.

check this https://github.com/behrad/aedes-cached-persistence/blob/DedupQlobber/index.js#L45 @davedoesdev

**Comment by davedoesdev** [comment]
Ah yes I missed https://github.com/behrad/aedes-cached-persistence/blob/DedupQlobber/index.js#L3

Match performance will be interesting. QlobberDedup has to add stuff to a Set and then split the strings, QlobberSub adds to a Map.

https://github.com/chad3814/node-hashtable is worth a go.

**Comment by davedoesdev** [comment]
@behrad version using node-hashtable here: https://github.com/davedoesdev/aedes-persistence/blob/node-hashtable/qlobber-sub.js

**Comment by davedoesdev** [comment]
Another option is using Object properties

**Comment by davedoesdev** [comment]
@behrad version using es6-native-map here: https://github.com/davedoesdev/aedes-persistence/blob/es6-native-map/qlobber-sub.js

**Comment by davedoesdev** [comment]
@behrad if you could see how those perform that would be great.
Also, they may impact match performance too.

**Comment by behrad** [comment]
I should port your examples inside my own fork of `aedes-cached-persistence` first, and then create a `redis-persistence` that is working with that specific aedes-cached-persistence.
Then it will be ready to compare them in real world.
Will let you know...

**Comment by behrad** [comment]
> Also, they may impact match performance too

I haven't yet tested match performance...
Can you elaborate on this more please @davedoesdev ?
Does matching time order change that much? 
Does matching time depend on the size of matches? (wildcard matching time)

My first citizen issue was, high CONNECT-restore-SUB ratio in production is generating a tons of calls to `.add` and we should not add duplicates. 
So there should be either a performance `dedup_add` OR a very fast `test_then_add` combination. Remembering that the wildcards are important here! since there exists some `.add`s to wildcard topics with millions records which `dedup` OR `test` should work on!

**Comment by behrad** [comment]
Comparing @davedoesdev cached-persistence using QlosQlobber
VS my cached-persistence using DedupQlobber:

| fork | offlineClients | subscriptions | Startup Time | Heap/RSS |
-------|------|-------|------|-------|
|https://github.com/behrad/aedes-cached-persistence/tree/DedupQlobber|1221271|3289187|147096.027ms|4436.0MB/5073.5MB
|https://github.com/davedoesdev/aedes-cached-persistence|1221334|3289353|163276.166ms|5287.1MB/6074.3MB

1 GB increased mem usage for a million clients.
Can you port you persistence alternate branches to cached-persistence, or help me how to use them in my cached-persistence so that I can test node-hashtable, ...?

**Comment by davedoesdev** [comment]
Just use es6-native-map and replace Map. However I found it much slower.

Also I don't think your QlobberDedup solution complies with the MQTT spec
because it does replace subscriptions with same topic but different SoD.

**Comment by mcollina** [comment]
Do we have some conclusion? It has been hard to follow. It seems the best path would be:
Merge this and the relative PRs in aedes-persistence and aedes-persistence-redis.
Measure and re-evaluate if we can get a faster solution, as this improves the current baselevel performance and it is still correct according to the spec.

**Comment by behrad** [comment]
After running https://github.com/davedoesdev/aedes-cached-persistence for an hour, I'm seeing some glitches in publish response time from my client. 
Don't know why, but switching back to my qlobberDedup branch fade that issue away. May be there's something going bad inside `match` with my size of data.. !!?

**Comment by behrad** [comment]
I was running https://github.com/behrad/aedes-persistence-redis/blob/with_qos_qlobber/persistence.js which is using https://github.com/behrad/aedes-cached-persistence/blob/QosQlobber/index.js 

Can you confirm that code was ok @davedoesdev ?

**Comment by behrad** [comment]
> Do we have some conclusion?

I think https://github.com/mcollina/aedes-persistence/pull/18 can be merged for starters...

**Comment by davedoesdev** [comment]
If a new subscription comes in with QoS 0 then we have to remove any
existing off line (QoS) subscriptions to the same topic. Spec says we must
support replacing subscriptions with same topic but different QoS.

On 7 Aug 2017 4:34 p.m., "Behrad" <notifications@github.com> wrote:

*@behrad* requested changes on this pull request.
------------------------------

In index.js
<https://github.com/mcollina/aedes-cached-persistence/pull/11#discussion_r131629657>
:

> @@ -93,7 +84,6 @@ CachedPersistence.prototype._addedSubscriptions = function (client, subs, cb) {
     }
   })

-  subs = subs.filter(qosGreaterThanOne)

Why have you removed this @davedoesdev <https://github.com/davedoesdev> ?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/mcollina/aedes-cached-persistence/pull/11#pullrequestreview-54626790>,
or mute the thread

**Comment by davedoesdev** [comment]
Yes that looks like it pulls in the latest code. What glitches are you
seeing? Correctness or performance?

**Comment by behrad** [comment]
> Correctness or performance?

Performance, it was no good. I rolled out to my own fork for now, but I will check to see if I can find out what is happening

**Comment by davedoesdev** [comment]
Very interesting. Can you repro outside of your system with a test program?
In the comparisons I posted QlobberSub matched much faster than
QlobberDedup.

**Comment by behrad** [comment]
When https://github.com/davedoesdev/aedes-persistence/blob/master/qlobber-sub.js#L26 is called @davedoesdev ?

**Comment by davedoesdev** [comment]
@behrad as part of matching (when collating the results).

**Comment by behrad** [comment]
> @behrad as part of matching (when collating the results)

A good suspect.

Let's alter https://github.com/mcollina/aedes-persistence/pull/18#issuecomment-320470156 to a new test that is like this in 3 phases:

```js
PHASE I: populating an empty Qlobber (adds are new)
i = 1 .. 1000000
  client = 'blah'+i
  x = random of 1 .. 3
  // each client is subscribed to 6 topics: 3 per client, 3 public
  qlobber.add(client, client + '/default')
  qlobber.add(client, client + '/12345')
  qlobber.add(client, client + '/test')
  qlobber.add(client, 'public/default')
  qlobber.add(client, 'public/test')
  qlobber.add(client, 'public/' + x)

PHASE II: randomly adding old clients to populated Qlobber (adds are mainly deduped)
// we can simply re-run 1..1000 of the above code on the same qlobber object

PHASE III: calling 10000 matches on per client topics

PHASE III-2: calling 1000 match on public topics
```

**Comment by davedoesdev** [comment]
@behrad thanks. Are you running Node with expanded heap?

**Comment by behrad** [comment]
I've extend your test script to all 5 cases, interesting results... 
I'll post in a few minutes!

**Comment by behrad** [comment]
> Node with expanded heap?

Do you mean `--max-old-space-size`? in production: YES!

**Comment by behrad** [comment]
This is the extended bench script which,
1. Adds 300K clientsId each with 3 single + 3 shared topics to an empty Qlobber (simulates startup setup time)
2. Re-runs add, to test how de-duping performs (simulates clean=false client connects)
3. Remove a little clients, to simulate unsubscribes
4. Matches against 100000 single topics (publishing to 1 client)
5. Matches against 10 shared topics (publishing to many clients)

This are most important scenarios in aedes (did I miss one?). In production, we saw 2 & 4 are the hottest, then 1 & 5.

```js
var QlobberSub = require('../aedes/qlobber-sub');
var QlobberOpts = {
  wildcard_one: '+',
  wildcard_some: '#',
  separator: '/'
}
var matcher = new QlobberSub(QlobberOpts);

function add () {
  var add_start = new Date();
  for (var i = 0; i < 300000; i += 1)
  {
    var x = Math.floor(Math.random() * 3 + 1)
    var clientId = 'someClientId/' + i;
    var topic = 'a/b/c/d/' + i;
    var topic2 = 'a/b/c/d/def' + i;
    var topic3 = 'a/b/c/d/id' + i;
    var topic4 = 'a/b/c/public/test';
    var topic5 = 'a/b/c/public/all';
    var topic6 = 'a/b/c/public/'+x;
    matcher.add(topic, { clientId: clientId, topic: topic, qos: 1 });
    matcher.add(topic2, { clientId: clientId, topic: topic2, qos: 1 });
    matcher.add(topic3, { clientId: clientId, topic: topic3, qos: 1 });
    matcher.add(topic4, { clientId: clientId, topic: topic4, qos: 1 });
    matcher.add(topic5, { clientId: clientId, topic: topic5, qos: 1 });
    matcher.add(topic6, { clientId: clientId, topic: topic6, qos: 1 });
  }
  var add_end = new Date();
  console.log('add', add_end.getTime() - add_start.getTime());
}

function remove () {
  var add_start = new Date();

  for (var i = 0; i < 10; i += 1)
  {
    var x = Math.floor(Math.random() * 3 + 1)
    var clientId = 'someClientId/' + i;
    var topic = 'a/b/c/d/' + i;
    var topic2 = 'a/b/c/d/def' + i;
    var topic3 = 'a/b/c/d/id' + i;
    var topic4 = 'a/b/c/public/test';
    var topic5 = 'a/b/c/public/all';
    var topic6 = 'a/b/c/public/'+x;
    matcher.remove(topic, { clientId: clientId, topic: topic, qos: 1 });
    matcher.remove(topic2, { clientId: clientId, topic: topic2, qos: 1 });
    matcher.remove(topic3, { clientId: clientId, topic: topic3, qos: 1 });
    matcher.remove(topic4, { clientId: clientId, topic: topic4, qos: 1 });
    matcher.remove(topic5, { clientId: clientId, topic: topic5, qos: 1 });
    matcher.remove(topic6, { clientId: clientId, topic: topic6, qos: 1 });
  }

  var add_end = new Date();

  console.log('remove', add_end.getTime() - add_start.getTime());
}

function match () {
  var match_start = new Date();
  for (var i = 0; i < 100000; i += 1)
  {
    var topic = 'a/b/c/d/' + i;
    matcher.match(topic);
  }
  var match_end = new Date();
  console.log('match', match_end.getTime() - match_start.getTime());
}

function matchMany () {
  var match_start = new Date();

  for (var i = 0; i < 10; i += 1)
  {
    var topic = 'a/b/c/public/test';

    console.log(matcher.match(topic).length);
  }

  var match_end = new Date();

  console.log('matchMany', match_end.getTime() - match_start.getTime());
}

function memUsage () {
  const mem = process.memoryUsage();
  const rss = (mem.rss / 1024 / 1024).toFixed(1);
  const heap = (mem.heapUsed / 1024 / 1024).toFixed(1);
  console.log(`Heap=${heap}MB, RSS=${rss}MB`);
}

memUsage();

add()

memUsage();

add()

memUsage();

remove()

memUsage();

match()

matchMany()
```

I've created 5 instances of the same script:
1. QosQlobber from aedes/qlobber-sub
2. DedupQlobber from QlobberDedup
3. ObjectQlobber from branch object-properties
4. HashtableQlobber from branch hashtable
5. NativeMapQlobber from branch native-map
all from @davedoesdev samples.

And now the results:

|type|add|dup-add|remove|match|matchMany|Memory|
|-----|-----|-----|-----|-----|-----|-----|
|QosQlobber | 5792 | 2713 | 1 | 223 | 2450 | 1046mb |
DedupQlobber | 4533 | 5127 | 1 | 706 | 2 | 858mb
ObjectQlobber | 8010 | 2975 | 2242 | 285 | 1921 | 916mb
HashtableQlobber | 10273 | 7598 | 2 | 312 | 4128 | 489mb
NativeMapQlobber | 13171 | 8799 | 1 | 1268 | 20574 | 515mb

This now explains why aedes using QosQlobber is hiccuped: 
https://github.com/behrad/aedes-persistence-redis/blob/master/persistence.js#L151
that is I am using `.match` in removeSubscriptions to check that if to be removed client is the last one subscribed to the topic, and remove the topic from topics list if so! So incoming UNSUBSCRIBES (remember 6 topics, 3 is shared) was leading to matchMany :)

Do you have any Idea for a better implementation https://github.com/behrad/aedes-persistence-redis/blob/master/persistence.js#L151?

**Comment by davedoesdev** [comment]
@behrad thanks. Let me absorb it for a while and get back with an answer. I'll put the bench scripts into the qlobber repo so we can keep them.

**Comment by davedoesdev** [comment]
@behrad have you tried https://github.com/mcollina/aedes-persistence-redis/pull/42 ?

**Comment by behrad** [comment]
> have you tried mcollina/aedes-persistence-redis#42 ?

replacing .match with .test? I don't think that works like https://github.com/mcollina/aedes-persistence-redis/pull/42/files#diff-8cebf2b673cb6d5ae4ccb42c33669c19R181, Since we don't want to check for existence of a client subscription, but also to check if anybody is still subscribed to that topic.

We can also drop `subs` key from redis persistence. What dya think @mcollina ?

**Comment by mcollina** [comment]
> https://github.com/davedoesdev/aedes-persistence-redis/blob/36f0816f927b7086fb96a9ccd9e8748813773a4a/persistence.js#L250

@behrad if we can provide the countOffline in another way, sure. If we have this info in aedes-cached-persistence, we can even provide a generic implementation for that, and remove some code here and in mongodb.

**Comment by behrad** [comment]
> if we can provide the countOffline in another way, sure. If we have this info in aedes-cached-persistence, we can even provide a generic implementation for that, and remove some code here and in mongodb.

https://github.com/davedoesdev/qlobber/issues/12

**Comment by davedoesdev** [comment]
We can either add a count method or modify test_values so it can just take
topic

**Comment by davedoesdev** [comment]
First thing to do is add the missing test case which didn't catch testing
per client rather than overall.



On 8 Aug 2017 12:40, "David Halls" <dahalls@gmail.com> wrote:

We can either add a count method or modify test_values so it can just take
topic

**Comment by davedoesdev** [comment]
@behrad I've updated by benchmark scripts to match yours (I think): https://github.com/davedoesdev/qlobber/tree/master/aedes/bench

Please check them over to check they do the same. I added the parsing to the QlobberDedup match results to make them produce the same results and keep them comparing the same thing.

**Comment by davedoesdev** [comment]
For QlobberDedup I get:

```
add_to_qlobber: 3567
add_to_qlobber: 4195
remove_from_qlobber: 1
match_client_topics: 306
match_public_topics: 4951
```

For QlobberSub:

```
add_to_qlobber: 4399
add_to_qlobber: 2746
remove_from_qlobber: 1
match_client_topics: 195
match_public_topics: 1812
```

- which matches my previous results (QlobberSub is slower to add, quicker to match).

**Comment by behrad** [comment]
> which matches my previous results (QlobberSub is slower to add, quicker to match).

good thing is that QlobberSub is also quicker on deduping (second add)

All results looks OK to me. The only strange one is QlobberDedup's match_public_topics! Mine was under 1000

**Comment by davedoesdev** [comment]
Could you try adding string split and parse the qos?
Be good to get our results looking similar first.

**Comment by behrad** [comment]
This now is comparable. Two issues we should address now:

1. A total topic `count` to implement offline count in persistence.

2. Space optimization inside QlobberSub @davedoesdev ?

**Comment by davedoesdev** [comment]
Yep. We still need to be able to tell when topic has no subscriptions?

**Comment by davedoesdev** [comment]
@behrad  I've added memory usage stats to my bench programs (https://github.com/davedoesdev/qlobber/tree/master/aedes/bench), comparing usage before all the operations with usage after, and get similar results to you:

Dedup: 

```
heap: 654.6MiB rss: 1259.2MiB
```

Sub:

```
heap: 922.7MiB rss: 1215.7MiB
```

(I'm forcing `gc` before taking results).

**Comment by davedoesdev** [comment]
OK I've changed `QlobberSub` so it only stores one copy of the topic per map entry. This has reduced memory usage (and times):

```
add_to_qlobber: 3075ms
add_to_qlobber: 1850ms
remove_from_qlobber: 1ms
match_client_topics: 163ms
match_public_topics: 1275ms
heap: 567.8MiB rss: 804.3MiB
```

**Comment by davedoesdev** [comment]
I'll commit that first, check the aedes tests still work and then see if I can change the `test()` implementation so it's not per-client.

**Comment by davedoesdev** [comment]
@behrad how does https://github.com/behrad/aedes-persistence-redis/blob/master/persistence.js#L151 know that `client` is the last one subscribed? What if a different client is the last one subscribed to the topic? Also, what if another wildcard topic matches?

**Comment by davedoesdev** [comment]
Would something like this do?

```javascript
QlobberSub.prototype.test_values = function (existing, val)
{
  return (existing.topic === val.topic) &&
         (existing.clientMap.size === 1) &&
         existing.clientMap.has(val.clientId);
};
```

The first line checks it's the same topic (i.e. ignores wildcards that might match).
The second line checks it's the last subscription.
The third line checks it's the specified client which is the last one subscribed.

**Comment by davedoesdev** [comment]
If that's right, then a version using QlobberDedup should be doing the following instead of just checking match length equals 1.

```javascript
function last_subscribed(matcher, clientId, topic)
{
    var count = 0;
    var found = false;
    
    for (var m of matcher.match(topic))
    {   
        var parts = m.split(';');
        if (parts[1] === topic)
        {   
            count += 1; 
            if (count > 1)
            {   
                return false;
            }
            if (parts[0] === clientId)
            {   
                found = true;
            }
        }
    }
    
    return (count === 1) && found;
```

@behrad @mcollina is that right?

**Comment by behrad** [comment]
@davedoesdev that code only was aimed to be able to count number of offline subscribed topics, If qlobber can tell about the total number of topics inside, we are free to completely remove that code.

**Comment by davedoesdev** [comment]
@behrad yep, I'll see if I can add counting to QlobberSub.

File: test.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ function MyPersistence () {
 
   // copied from Memory
   this._retained = []
-  this._subscriptions = []
+  this._subscriptions = new Map()
```
Comment (behrad) [inline test.js]: @davedoesdev

Comment (davedoesdev) [inline test.js]: This is used for the tests only and is copied over from Memory persistence in aedes-persistence. A bunch of methods are copied over from Memory persistence too, including subscriptionsByClient which needs all the subscriptions, including qos 0 ones which aren't stored in the QlobberSub.

Comment (davedoesdev) [inline test.js]: Each persistence will choose its own way of storing subscriptions. aedes-persistence-redis, for example, stores it in a Redis hash key.

**Comment by behrad** [comment]
> This is used for the tests only and is copied over from Memory persistence in aedes-persistence.

cached-persistence is a kind of abstract persistence, sitting in the middle of bare bone persistence and concrete store persistences like redis, mogo, etc. 
We should remove this, otherwise we are replicating all subscriptions in memory in all child persistences. and that's not acceptable memory wise. Fix me please @mcollina

**Comment by mcollina** [comment]
What should I fix? Can't it be done in a PR?

**Comment by davedoesdev** [comment]
@behrad where does cached-persistence store subscriptions in memory?

**Comment by behrad** [comment]
> where does cached-persistence store subscriptions in memory?

My mistake guys! I was totally missing that I was reading `test.js` so `this._subscriptions = new Map()` is inside tests! Pordon me @davedoesdev 

I'm ok with this @mcollina

**Comment by davedoesdev** [comment]
@behrad no problem.
Could we resolve https://github.com/mcollina/aedes-persistence/issues/20 before merging this PR? It seems like a trivial issue but has implications for other persistences.
